src/VisualStudio/Core/Def/Implementation/Workspace/SourceGeneratedFileManager.cs

```diff
@@ -0,0 +1,422 @@
+
+        /// <summary>
+        /// The temporary directory that we'll create file names under to act as a prefix we can later recognize and use.
+        /// </summary>
+        private readonly string _temporaryDirectory;
+
+        /// <summary>
+        /// Map of currently open generated files; the key is the generated full file path.
+        /// </summary>
+        private readonly Dictionary<string, OpenSourceGeneratedFile> _openFiles = new Dictionary<string, OpenSourceGeneratedFile>();
+        private readonly VisualStudioWorkspace _visualStudioWorkspace;
+
+        [ImportingConstructor]
+        [Obsolete(MefConstruction.ImportingConstructorMessage, error: true)]
+        public SourceGeneratedFileManager(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            IThreadingContext threadingContext,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService,
+            ITextDocumentFactoryService textDocumentFactoryService,
+            VisualStudioWorkspace visualStudioWorkspace)
+        {
+            _serviceProvider = serviceProvider;
+            _threadingContext = threadingContext;
+            _textDocumentFactoryService = textDocumentFactoryService;
+            _temporaryDirectory = Path.Combine(Path.GetTempPath(), "VisualStudioSourceGeneratedDocuments");
+            _visualStudioWorkspace = visualStudioWorkspace;
+
+            Directory.CreateDirectory(_temporaryDirectory);
+
+            _runningDocumentTable = _serviceProvider.GetService<SVsRunningDocumentTable, IVsRunningDocumentTable>();
+            _runningDocumentTableEventTracker = new RunningDocumentTableEventTracker(
+                threadingContext,
+                editorAdaptersFactoryService,
+                _runningDocumentTable,
+                this);
+        }
+
+        public void NavigateToSourceGeneratedFile(Project project, ISourceGenerator generator, string generatedSourceHintName, TextSpan sourceSpan)
+        {
+            // We will create an file name to represent this generated file; the Visual Studio shell APIs imply you can use a URI,
+            // but most URIs are blocked other than file:// and http://; they also get extra handling to attempt to download the file so
+            // those aren't really usable anyways.
+
+            // The file name we generate here is chosen to match the compiler's choice, so the debugger can recognize the files should match.
+            // We use the ProjectId as a parent directory so we can extract that out later too.
+            var generatorType = generator.GetType();
+            var temporaryFilePath = Path.Combine(_temporaryDirectory, $"{project.Id.Id}\\{generatorType.Module.ModuleVersionId}_{generatorType.FullName}_{generatedSourceHintName}");
+            Directory.CreateDirectory(Path.GetDirectoryName(temporaryFilePath));
+            File.WriteAllText(temporaryFilePath, "");
```

Comment:
💡 I would prefer to write a temporary random file, and then delete it as soon as it opens. This addresses concerns about both the length of the name, and the fact that users can control the name.

Reply:
I'll look at the deleting, but the name of the file is critical in that it must line up with what the compiler writes into the PDB. I'll clarify that that's not changeable, but @chsienki and I are discussing making another change there.

Reply:
#47212 filed to track investigation here.